net/graphsync_fetcher.go

```
@@ -78,38 +88,79 @@ const recursionMultiplier = 4
 // go-filecoin migrates to the same IPLD library used by go-graphsync (go-ipld-prime)
 //
 // See: https://github.com/filecoin-project/go-filecoin/issues/3175
-func (gsf *GraphSyncFetcher) FetchTipSets(ctx context.Context, tsKey types.TipSetKey, from peer.ID, done func(types.TipSet) (bool, error)) ([]types.TipSet, error) {
-	cids := tsKey.ToSlice()
-	err := gsf.fetchBlocks(ctx, cids, from)
+func (gsf *GraphSyncFetcher) FetchTipSets(ctx context.Context, tsKey types.TipSetKey, initialPeer peer.ID, done func(types.TipSet) (bool, error)) ([]types.TipSet, error) {
+	rpf := newRequestPeerFinder(gsf.peerTracker, initialPeer)
+
+	// fetch initial tipset
+	startingTipset, err := gsf.fetchFirstTipset(ctx, tsKey, rpf)
 	if err != nil {
 		return nil, err
 	}
-	startingTipset, err := gsf.loadTipsetFromCids(ctx, cids)
-	if err != nil {
-		return nil, err
+
+	// fetch remaining tipsets recursively
+	return gsf.fetchRemainingTipsets(ctx, startingTipset, rpf, done)
+}
+
+func (gsf *GraphSyncFetcher) fetchFirstTipset(ctx context.Context, tsKey types.TipSetKey, rpf *requestPeerFinder) (types.TipSet, error) {
+	cids := tsKey.ToSlice()
+	for {
+		err := gsf.fetchBlocks(ctx, cids, rpf.CurrentPeer())
+		if err == nil {
+			break
+		}
+		// something went wrong in a graphsync request, but we want to keep trying other peers, so
+		// just log error
+		logGraphsyncFetcher.Infof("Error occurred during Graphsync request: %s", err)
+		cids, err = gsf.missingCids(cids)
+		if err != nil {
+			return types.UndefTipSet, err
+		}
+		err = rpf.FindNextPeer()
+		if err != nil {
+			return types.UndefTipSet, fmt.Errorf("Failed fetching tipset: %s", tsKey.String())
```

The "no more peers" error is swallowed, but should be wrapped